Add custom audio renderer for iPhone earpiece and only render joined participants

toger5 opened the pull request:
This does two things:
 - It will filter playback tracks by matrixRTC members (only actual participants will ever be audible) https://github.com/element-hq/element-call/issues/2927
 - It will add a new "virtual device" option for ios to enable earpice speaker
 
Fixes https://github.com/element-hq/element-call/issues/2927

Half-Shot commented on src/mediapipe/imageSegmenter/audioNativeRenderer/MatrixAudioRenderer.tsx:
```diff
@@ -0,0 +1,144 @@
+  /** Does an additional gain control on all audio tracks rendered by this component. This does not interfere with the
+   * `setVolume` of the track itself.
+   */
+  volumeFactor?: number;
```
Bit confused by this. Is the volumeFactor cumulative or just the default?

I.e. is it

1. `volumeFactor * trackVolumeFactor` for each track?
2. `trackVolumeFactor ||  volumeFactor` for each track?

toger5 replied:
its
`track.connect(gainNode).connect(panNode).connect(gainNode2)`
where:
`gainNode`: is the configured by the volumeFactor and `gainNode2` is configured by setVolume on the RemoteAudioTrack
So it think 1. fits better but then gain is not linear i think. so its similar but different.

Half-Shot replied:
alright, that sounds like it would match what I'd expect the buttons to do :)

Half-Shot commented on src/mediapipe/imageSegmenter/audioNativeRenderer/MatrixAudioRenderer.tsx:
```diff
@@ -0,0 +1,144 @@
+}
+
+/**
+ * The `RoomAudioRenderer` component is a drop-in solution for adding audio to your LiveKit app.
```
This seems like the description of the parent component?

Half-Shot commented on src/mediapipe/imageSegmenter/audioNativeRenderer/MatrixAudioRenderer.tsx:
```diff
@@ -0,0 +1,144 @@
+      }
+
+      // Remove nodes from trackNodes map.
+      trackNodes.current.forEach((_, identity) => {
```
Do we need to do any cleanup of the audioContext when removing tracks? Can they end up added twice or something?

Half-Shot commented on src/mediapipe/imageSegmenter/audioNativeRenderer/MatrixAudioRenderer.tsx:
```diff
@@ -0,0 +1,144 @@
+  }, [balance, tracks, volumeFactor]);
+
+  return (
+    <div style={{ display: "none" }}>
```
ooi, do the AudoTracks' render anything? We need to display: none them?

toger5 replied:
I think this was just to group them into one collapsable section?

robintown replied:
A comment that this `<div>` exists for dev experience wouldn't hurt

Half-Shot commented on src/settings/SettingsModal.tsx:
```diff
@@ -121,6 +122,16 @@ export const SettingsModal: FC<Props> = ({
             numberedLabel={(n) => t("settings.devices.speaker_numbered", { n })}
           />
+          <FieldRow>
+            <InputField
```
This is more of a product nit, but I'm surprised we don't just expose a pan setting like most volume controls do.

toger5 replied:
Yes we definitly should. I am not sure what solution I should use however. android does expose this autmatically so we dont need the stereo pan for it.
So we could make this a url flag. Or maybe there is an obvious mobile safari browser hint we can consume and expose the option then.

Half-Shot reviewed (dismissed):
Just had some questions. Implementation looks sane

robintown commented on src/useAudioContext.tsx:
```diff
@@ -105,23 +111,30 @@ export function useAudioContext<S extends string>(
       });
     }
-  }, [audioContext, devices]);
+  }, [audioContext, audioOutput]);
```
```suggestion
  }, [audioContext, audioOutput.selectedId]);
```

Hopefully React is cool with this?

robintown commented on src/useAudioContext.test.tsx:
```diff
@@ -39,94 +39,156 @@ const TestComponent: FC = () => {
   public static testContext: MockAudioContext;
 
   public constructor() {
+    if (MockAudioContext.testContext) {
+      this.gain = MockAudioContext.testContext.gain;
+      this.pan = MockAudioContext.testContext.pan;
+      this.setSinkId = MockAudioContext.testContext.setSinkId;
+      this.decodeAudioData = MockAudioContext.testContext.decodeAudioData;
+      this.createBufferSource = MockAudioContext.testContext.createBufferSource;
+      this.createGain = MockAudioContext.testContext.createGain;
+      this.createStereoPanner = MockAudioContext.testContext.createStereoPanner;
+      this.close = MockAudioContext.testContext.close;
+    }
```
This is a bit awkward, I take it you're trying to have repeated calls to `new MockAudioContext` return "the same value" essentially. I think you could achieve this more easily by using a simple object + function instead of the ES6 `class` syntax:

```typescript
const testContext = { gain: ..., pan: ..., /* et cetera */ }
function MockAudioContext() { return testContext }

const ctx1 = new MockAudioContext()
const ctx2 = new MockAudioContext()
expect(ctx1).toBe(ctx2) // this now works
```

toger5 replied:
I ended up to keep testContext a class instance instead of a struct/object.
Reason being that we need the types in the test to `expect` the mocked values.
Those types otherwise wuold have need to been defined with an additional interface.
This way we get it automatically.

robintown commented on src/settings/DeveloperSettingsTab.tsx:
```diff
@@ -175,6 +179,20 @@ export const DeveloperSettingsTab: FC<Props> = ({ client, livekitRoom }) => {
+          id="alwaysShowIphoneEarpiece"
+          type="checkbox"
+          label={t("developer_mode.always_show_iphone_earpiece")}
+          checked={!!alwaysShowIphoneEarpiece}
```
```suggestion
          checked={alwaysShowIphoneEarpiece}
```

robintown commented on src/mediapipe/imageSegmenter/audioNativeRenderer/MatrixAudioRenderer.tsx:
Why has this ended up in the MediaPipe folder?

toger5 replied:
It seemed to be most related.

toger5 replied:
I move it to the livekit folder.

robintown commented on src/mediapipe/imageSegmenter/audioNativeRenderer/MatrixAudioRenderer.tsx:
```diff
@@ -0,0 +1,215 @@
+   * If set to `true`, mutes all audio tracks rendered by the component.
+   * @remarks
+   * If set to `true`, the server will stop sending audio track data to the client.
+   * @alpha
```
Why `@alpha`? Is this expected to be unstable?

robintown commented on src/mediapipe/imageSegmenter/audioNativeRenderer/MatrixAudioRenderer.tsx:
````diff
@@ -0,0 +1,215 @@
+ * </LiveKitRoom>
+ * ```
+ * @public
+ */
````
This doc comment looks copied & pasted / inaccurate now that this component is more custom

robintown commented on src/mediapipe/imageSegmenter/audioNativeRenderer/MatrixAudioRenderer.tsx:
```diff
@@ -0,0 +1,215 @@
+  muted,
+}: MatrixAudioRendererProps): ReactNode {
+  const validIdentities = useMemo(
+    () => members?.map((member) => member.sender + ":" + member.deviceId),
```
```suggestion
    () => members?.map((member) => `${member.sender}:${member.deviceId}`),
```

Templates are more performant whenever there are more than 2 parts to the string, IIRC

robintown commented on src/mediapipe/imageSegmenter/audioNativeRenderer/MatrixAudioRenderer.tsx:
```diff
@@ -0,0 +1,215 @@
+      onlySubscribed: true,
+    },
+  ).filter((ref) => {
+    const isValid = validIdentities?.includes(ref.participant.identity);
```
Perhaps it's appropriate for `validIdentities` to be a `Set`, if this is what we're using it for

robintown commented on src/mediapipe/imageSegmenter/audioNativeRenderer/MatrixAudioRenderer.tsx:
```diff
@@ -0,0 +1,215 @@
+    },
+  ).filter((ref) => {
+    const isValid = validIdentities?.includes(ref.participant.identity);
+    if (!isValid) logInvalid(ref.participant.identity, validIdentities);
```
This looks like it could cause a lot of repeated logs for the same users, as it runs on every update to the list of tracks. Is there an easy mitigation here to only log once per participant?

It also looks like this line could end up warning that we're not rendering the local user's audio, since it comes before the `!ref.participant.isLocal` branch, even though that'd be a non-issue.

toger5 replied:
Our local participation also should not be invalid rigth?
Even though it would not break the call the log still might be helpful, that we connected to lk before our memberhsip event was set.
Since we do not have local echos this might be a fairly commen case however...

robintown replied:
I'm thinking of the case where my client fails to refresh the delayed event due to poor homeserver performance while I remain connected to the SFU, in the middle of the call. That happens to me every few days in our stand-ups.

> Even though it would not break the call the log still might be helpful, that we connected to lk before our memberhsip event was set.

Potentially, though warning that we're not rendering our own audio is still a misleading thing to say in this case

toger5 replied:
I added a not is local check.

robintown commented on src/mediapipe/imageSegmenter/audioNativeRenderer/MatrixAudioRenderer.tsx:
```diff
@@ -0,0 +1,215 @@
+   * This additional volume factor is applied to all audio tracks from livekit.
+   *
+   * @remark It does not interfere with the `setVolume` of the track itself.
+   * The seperta `volume` property will override/conflict with what is set using `track.setVolume`
```
```suggestion
   * The seperate `volume` property will override/conflict with what is set using `track.setVolume`
```

robintown commented on src/mediapipe/imageSegmenter/audioNativeRenderer/MatrixAudioRenderer.tsx:
```diff
@@ -0,0 +1,215 @@
+   */
+  volumeFactor: number;
+  /**
+   * The left right stereo pan of the audio tracks.
```
```suggestion
   * The left-right stereo pan of the audio tracks.
```

robintown commented on src/mediapipe/imageSegmenter/audioNativeRenderer/MatrixAudioRenderer.tsx:
```diff
@@ -0,0 +1,215 @@
+  // the actual `audioContext` is stored in `audioContextRef` to minimize reallocations.
+  // whenever we notice the audioContext is active but should not: `!!audioContext !== shouldUseAudioContext`
+  // we remount the audio tracks using a true -> false -> true transition for the `trackReady` state.
+  const audioContextRef = useRef(new AudioContext());
```
This will create a new AudioContext on each render, which is dangerous: you need to at least call the `close()` method.

```suggestion
  const audioContext = useMemo(() => new AudioContext(), []);
  useEffect(() => () => audioContext.close(), [audioContext]);
```

But in general we've seen this cause big resource leaks in the past, back when walkie-talkie mode was a thing. Is there some way that we could just create a single AudioContext in the global scope and use that across the entire app? (that is, use the same context in `useAudioContext` too?) Given that an accumulation of AudioContexts can cause stuttering audio, I think we should consider them expensive.

toger5 replied:
Are you sure it would create one each render.
Or do you mean a new one on each creation of the compoenent?

But that will be the same for the memo version.

I am not sure we can have the same context with different destinations?
Now they are all just connected to one html element.
Lk also provides a way to pass a single mixed audio context to the lk room.
But this would be an entirely differnt architecture and I am not sure we would be able to enforece the correct remounting ourder as easily when switching from audio context to not using audio context.

I would postpone a central audioContext refactor once we have a decision if locking ios should mute audio or switch back to native html audio?

robintown commented on src/mediapipe/imageSegmenter/audioNativeRenderer/MatrixAudioRenderer.tsx:
```diff
@@ -0,0 +1,215 @@
+  }, [stereoPan]);
+  useEffect(() => {
+    // *4 to balance the transition from audio context to normal audio playback.
+    // probably needed due to gain bahving differently than el.volume
```
```suggestion
    // probably needed due to gain behaving differently than el.volume
```

robintown commented on src/mediapipe/imageSegmenter/audioNativeRenderer/MatrixAudioRenderer.tsx:
```diff
@@ -0,0 +1,215 @@
+  // we remount the audio tracks using a true -> false -> true transition for the `trackReady` state.
+  const audioContextRef = useRef(new AudioContext());
+
+  const gainRef = useRef(audioContextRef.current.createGain());
+  const panRef = useRef(audioContextRef.current.createStereoPanner());
```
These should use `useMemo` too

toger5 replied:
This would not make a difference except allowing us to call `.disconnect()`?

robintown commented on src/mediapipe/imageSegmenter/audioNativeRenderer/MatrixAudioRenderer.tsx:
```diff
@@ -0,0 +1,215 @@
+  const [trackReady, setTrackReady] = useState(false);
+
+  const [isVisible, setIsVisible] = useState(true);
+  useEventTarget(document, "visibilitychange", () =>
+    setIsVisible(!document.hidden),
+  );
```
```suggestion
  useEventTarget(document, "visibilitychange", useCallback(() =>
    setIsVisible(!document.hidden),
  []));
```

toger5 replied:
Why is that needed? It seems like bette hygene but what would happen wihtou use callback?

robintown commented on src/mediapipe/imageSegmenter/audioNativeRenderer/MatrixAudioRenderer.tsx:
```diff
@@ -0,0 +1,215 @@
+  // (adding the audio context when already mounted did not work outside strict mode)
+  const [trackReady, setTrackReady] = useState(false);
+
+  const [isVisible, setIsVisible] = useState(true);
```
```suggestion
  const [isVisible, setIsVisible] = useState(!document.hidden);
```

to be safe?

robintown commented on src/mediapipe/imageSegmenter/audioNativeRenderer/MatrixAudioRenderer.tsx:
```diff
@@ -0,0 +1,215 @@
+  const volume = useMemo(
+    () => (audioContextEnabled.current ? props.volume : volumeFactor),
+    [audioContextEnabled, props.volume, volumeFactor],
+  );
```
This value is not guaranteed to be reactive with respect to `audioContextEnabled.current`, since that's a ref.

robintown commented on src/mediapipe/imageSegmenter/audioNativeRenderer/MatrixAudioRenderer.tsx:
```diff
@@ -0,0 +1,215 @@
+    setIsVisible(!document.hidden),
+  );
+
+  useEffect(() => {
+    // Once sth changes (stereoPan or isVisible) so that we change if we use the audioContext.
+    // We will disable the audio track.
+    // this will unmount the AudioTrack component.
+    // it will get remounted once the track is updated with the new audioContext configurations.
+    const shouldUseAudioContext = isVisible && stereoPan !== 0;
+    if (audioContextEnabled.current !== shouldUseAudioContext) {
+      audioContextEnabled.current = shouldUseAudioContext;
+      setTrackReady(false);
+    }
+  }, [audioContextEnabled, audioContextRef, isVisible, stereoPan]);
+
+  useEffect(() => {
+    if (!trackRef || trackReady) return;
+    const track = trackRef.publication.track as RemoteAudioTrack;
+    track.setAudioContext(
+      audioContextEnabled.current ? audioContextRef.current : undefined,
+    );
+    track.setWebAudioPlugins(
+      audioContextEnabled.current ? [panRef.current, gainRef.current] : [],
+    );
+    setTrackReady(true);
+  }, [audioContextRef, trackReady, trackRef]);
```
I think you could get rid of the `audioContextEnabled` ref and instead react to changes in `shouldUseAudioContext`, like:

```typescript
  const shouldUseAudioContext = isVisible && stereoPan !== 0;
  useEffect(() => {
    // Once sth changes (stereoPan or isVisible) so that we change if we use the audioContext.
    // We will disable the audio track.
    // this will unmount the AudioTrack component.
    // it will get remounted once the track is updated with the new audioContext configurations.
    setTrackReady(false);
  }, [shouldUseAudioContext]);

  useEffect(() => {
    if (!trackRef || trackReady) return;
    const track = trackRef.publication.track as RemoteAudioTrack;
    track.setAudioContext(
      shouldUseAudioContext ? audioContextRef.current : undefined,
    );
    track.setWebAudioPlugins(
      shouldUseAudioContext ? [panRef.current, gainRef.current] : [],
    );
    setTrackReady(true);
  }, [audioContextRef, trackReady, trackRef, shouldUseAudioContext]);
```

or perhaps even:

```typescript
  const shouldUseAudioContext = isVisible && stereoPan !== 0;
  const [trackReady, setTrackReady] = useReactiveState(() => false, [shouldUseAudioContext]);

  useEffect(() => {
    if (!trackRef || trackReady) return;
    const track = trackRef.publication.track as RemoteAudioTrack;
    track.setAudioContext(
      shouldUseAudioContext ? audioContextRef.current : undefined,
    );
    track.setWebAudioPlugins(
      shouldUseAudioContext ? [panRef.current, gainRef.current] : [],
    );
    setTrackReady(true);
  }, [audioContextRef, trackReady, trackRef, shouldUseAudioContext]);
```

This could help with the previous comment about reactivity too.

toger5 replied:
Dont we also need to put shouldUseAudioContext into a memo?

robintown commented on src/livekit/MediaDevicesContext.tsx:
```diff
@@ -102,27 +111,39 @@ function useMediaDevice(
             // Create a virtual default audio output for browsers that don't have one.
             // Its device ID must be the empty string because that's what setSinkId
             // recognizes.
+            // we also create this if we do not have any available devices, so that
+            // we can use the default or the earpiece.
```
I'd be interested to know what platforms/conditions can cause there to be no available devices, because it was useful to have this `available.size` branch to hide the device list in settings during the loading state (while we're still awaiting the `enumerateDevices` call).

toger5 replied:
Thanks for the explanation.
The new change then seems to be the correct solution to me.

With the shortcoming, that we can already see default + earpice while we are still enumerating devices.
But it will remove the "fake default" once we are done enumerating.

To answer your question: I only tested android and it lists not output on enumerate devices.

robintown replied:
Nice, thanks for adding `available.size` back

robintown commented on src/livekit/MediaDevicesContext.tsx:
```diff
@@ -102,27 +111,39 @@ function useMediaDevice(
+              // the earpiece is not available as a device ID.
+              available = new Map([
+                ...available,
+                [EARPIECE_CONFIG_ID, { type: "name", name: "Earpiece" }],
```
This `"Earpiece"` string needs translation. Perhaps instead of using `type: "name"`, we could introduce a new `type: "earpiece"` variant to `DeviceLabel`.

toger5 replied:
That makes sense. Was thinking i could save the work and just reuse name but this ends up to be clearer.

robintown reviewed (commented):
Not fully through reviewing the last bits of the MediaDevicesContext changes but here's what I've got:

toger5 commented:
@robintown This has changed a bit since we cannot do the fallback to HTML5 audio because HTML5 audio volume control is broken on safari...
I rebased and squashed the changes to cleanup the PR and work on the playwright tests (thats why i wanted/needed to rebase)
I hope this is okay.
I only touched the `MatrixAudioRenderer` after your review. and it has changed enough so it makes sense to review the whole file again (it got much simpler)

toger5 committed:
Add custom audio renderer to only render joined participants & add ios earpice workaround

fix left right to match chromium + safari
(firefox is swapped)

earpice as setting

Simpler code and documentation
The doc explains, what this class actually does and why it is so complicated.

Signed-off-by: Timo K <toger5@hotmail.de>

use only one audioContext, remove (non working) standby fallback

toger5 committed:
Add tests

robintown commented on src/room/InCallView.test.tsx:
```diff
@@ -48,6 +44,8 @@ import {
 } from "../settings/settings";
 import { ReactionsSenderProvider } from "../reactions/useReactionsSender";
 import { useRoomEncryptionSystem } from "../e2ee/sharedKeyManagement";
+// import { testAudioContext } from "../useAudioContext.test";
```
```suggestion
```

robintown commented on src/settings/DeviceSelection.tsx:
```diff
@@ -22,30 +22,34 @@ import {
 
-  if (devices.available.size == 0) return null;
+  // There is no need to show the menu if there is no choice that can be made.
+  if (device.available.size == 1) return null;
```
```suggestion
  if (device.available.size <= 1) return null;
```

I suspect this will currently show a heading with no contents in the case of 0 devices (visible in Firefox while awaiting the `enumerateDevices` call)

toger5 replied:
I think you are right...

robintown commented on src/livekit/MatrixAudioRenderer.test.tsx:
```diff
@@ -0,0 +1,104 @@
+
+vi.mock("@livekit/components-react", async (importOriginal) => {
+  return {
+    ...(await importOriginal()), // this will only affect "foo" outside of the original module
```
Was "foo" meant to say something else?

robintown commented on src/livekit/MatrixAudioRenderer.tsx:
```diff
@@ -0,0 +1,201 @@
+    );
+  });
+
+  // This component is also (in addition to the "only play audio for connected members" logic above)
+  // to mimic earpice audio on iphones.
+  // The safari audio devices enumeration does not expose an earpice audio device.
+  // We alternatively use the audioContext pan node to only use one of the stereo channels.
+
+  // This component does get additionally complicated because of a safari bug.
+  // (see: https://bugs.webkit.org/show_bug.cgi?id=251532
+  // and the related issues: https://bugs.webkit.org/show_bug.cgi?id=237878
+  // and https://bugs.webkit.org/show_bug.cgi?id=231105)
+  //
+  // AudioContext gets stopped if the webview gets moved into the background.
+  // Once the phone is in standby audio playback will stop.
+  // So we can only use the pan trick only works is the phone is not in standby.
+  // If earpice mode is not used we do not use audioContext to allow standby playback.
+  // shouldUseAudioContext is set to false if stereoPan === 0 to allow standby bluetooth playback.
```
```suggestion
  // This component is also (in addition to the "only play audio for connected members" logic above)
  // responsible for mimicking earpiece audio on iPhones.
  // The Safari audio devices enumeration does not expose an earpiece audio device.
  // We alternatively use the audioContext pan node to only use one of the stereo channels.

  // This component does get additionally complicated because of a Safari bug.
  // (see: https://bugs.webkit.org/show_bug.cgi?id=251532
  // and the related issues: https://bugs.webkit.org/show_bug.cgi?id=237878
  // and https://bugs.webkit.org/show_bug.cgi?id=231105)
  //
  // AudioContext gets stopped if the webview gets moved into the background.
  // Once the phone is in standby audio playback will stop.
  // So we can only use the pan trick only works is the phone is not in standby.
  // If earpiece mode is not used we do not use audioContext to allow standby playback.
  // shouldUseAudioContext is set to false if stereoPan === 0 to allow standby bluetooth playback.
```

robintown commented on src/livekit/MatrixAudioRenderer.tsx:
```diff
@@ -0,0 +1,201 @@
+  const shouldUseAudioContext = stereoPan !== 0;
+
+  // initialize the potentially used audio context.
+  const audioContext = useMemo(() => new AudioContext(), []);
```
This will still result in an growing number of un-closed AudioContexts if you join and leave calls repeatedly. I think we do need to either move the AudioContext to a global/module level `const`, or else use the trick we discussed where you create the AudioContext in an effect, after all.

```typescript
const [audioContext, setAudioContext] = useState<AudioContext | undefined>(undefined)
useEffect(() => {
  const ctx = new AudioContext()
  setAudioContext(ctx)
  return () => ctx.close()
}, [])
```

(that one, just for reference)

toger5 replied:
I think I tried that and it did not work. But let me play around with that again.

robintown commented on src/livekit/MatrixAudioRenderer.tsx:
```diff
@@ -0,0 +1,201 @@
+    audioNodes.pan.pan.value = stereoPan;
+  }, [audioNodes.pan.pan, stereoPan]);
+  useEffect(() => {
+    // *4 to balance the transition from audio context to normal audio playback.
+    // probably needed due to gain behaving differently than el.volume
+    audioNodes.gain.gain.value = volumeFactor;
```
I don't see a `* 4` here

toger5 replied:
yes the *4 could get removed. Thanks the comment needs updating.

toger5 committed:
use optional audio context and effect to initiate it + review
